pre-docs/writing_tasks.md

```diff
@@ -234,6 +231,51 @@ Each task or plan name segment must begin with a lowercase letter and:
 - Namespace segments must match the following regular expression \A[a-z][a-z0-9_]*\Z
 - The file extension must not use the reserved extensions .md or .json.
 
+### Tasks with a single implementation
+
+A task can consist of a single executable with or without a corresponding metadata file. For
+instance, `./mysql/tasks/sql.rb` and `./mysql/tasks/sql.json`. In this case, no
+other `./mysql/tasks/sql.*` files can exist.
+
+### Tasks with multiple implementations
+
+A task can also have multiple implementation, with metadata that explains when
+to use each implementation. For instance, consider a module with the following
+files:
+
+```
+- tasks
+  - sql.sh
+  - sql.ps1
+  - sql.json
+```
+
+This task has two executables (`sql.sh` and `sql.ps1`) with a metadata file. The metadata file contains an `implementations` section:
+
+```json
+{
+  "implementations": [
+    {"name": "sql.sh", "requirements": ["shell"]},
+    {"name": "sql.ps1", "requirements": ["powershell"]}
+  ]
+}
+```
+
+Each implementations has a `name` and a list of `requirements`. The
```

Comment:
I think it's worth calling out that `name` can differ from the metadata file name, and that it's still restricted to valid task name patterns (I'm pretty sure Puppet will fail to find `my-impl.sh`).